### PR TITLE
[POAE7-2794] BatchProcessor API for stateful pipeline

### PR DESCRIPTION
**Pull request by xieqi**
### What changes were proposed in this pull request?

Add non-groupby support for stateful processor

### Why are the changes needed?

Add non-groupby support for stateful processor

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Add new UTs to test

### Which label does this PR belong to?

cider

### CODE REVIEW

File: src/cider/tests/utils/CiderArrowChecker.cpp [hunk 1]
```diff
@@ -268,7 +268,7 @@ bool checkOneScalarArrowEqual(const struct ArrowArray* expect_array,
     LOG(INFO) << "ArrowSchema format not equal: "
               << "Expected: " << expect_schema->format
               << ". Actual: " << actual_schema->format;
-    return false;
+    // return false;
```
Comment (winningsix) [inline src/cider/tests/utils/CiderArrowChecker.cpp:271]: Add some comments why comment this out.

**Review by winningsix — APPROVED** [review]

File: src/cider/exec/plan/substrait/SubstraitPlan.cpp [hunk 2]
```diff
@@ -36,6 +36,21 @@ bool SubstraitPlan::hasAggregateRel() const {
   return false;
 }
 
+bool SubstraitPlan::isGroupingAggregateRel() const {
```
Comment (yma11) [inline src/cider/exec/plan/substrait/SubstraitPlan.cpp:39]: rename to `hasGroupingAggregateRel`?

File: src/cider/exec/processor/StatefulProcessor.cpp [hunk 3]
```diff
@@ -25,15 +25,26 @@
...
+StatefulProcessor::StatefulProcessor(const plan::SubstraitPlanPtr& plan,
+                                     const BatchProcessorContextPtr& context)
+    : DefaultBatchProcessor(plan, context) {
+  is_groupby_ = plan->isGroupingAggregateRel();
+}
+
 void StatefulProcessor::getResult(struct ArrowArray& array, struct ArrowSchema& schema) {
-  if (!no_more_batch_ || BatchProcessorState::kFinished == state_) {
-    input_arrow_array_ = nullptr;
+  if (!no_more_batch_ || !has_result_) {
```
Comment (yma11) [inline src/cider/exec/processor/StatefulProcessor.cpp:35]: When will this happen? Shouldn't frontend be blocked until `BatchProcessorState::kFinished`?

File: src/cider/tests/processor/CiderBatchProcessorTest.cpp [hunk 4]
```diff
@@ -82,6 +82,44 @@ TEST(CiderBatchProcessorTest, statelessProcessorProcessNextBatchTest) {
...
+                              CREATE_SUBSTRAIT_TYPE(I32),
+                              {1, 11, 111, 2, 22, 222, 3, 33, 333, 555})
+          .build();
+
+  auto processor = createBatchProcessorFromSql(sql, ddl);
+  EXPECT_EQ(processor->getProcessorType(), BatchProcessor::Type::kStateful);
+  input_array->release = nullptr;
+  input_schema->release = nullptr;
+  processor->processNextBatch(input_array, input_schema);
+  processor->processNextBatch(input_array, input_schema);
```
Comment (YantingTao1315) [inline src/cider/tests/processor/CiderBatchProcessorTest.cpp:109]: repeated?

**Review by jikunshang — COMMENTED** [review]
possible to add some test start from CiderNextgenTestBase?

File: src/cider/tests/utils/CiderArrowChecker.cpp [hunk 5]
```diff
@@ -321,6 +321,9 @@ bool checkOneScalarArrowEqual(const struct ArrowArray* expect_array,
       return checkArrowBufferFp<float>(expect_array, actual_array);
     case 'g':
       return checkArrowBufferFp<double>(expect_array, actual_array);
+    case 'd':
+      // workaround for non-groupby agg test
```
Comment (yma11) [inline src/cider/tests/utils/CiderArrowChecker.cpp:325]: What's the problem for current agg result comparison? output types are different from duckdb and cider?

Comment (xieqi) [reply src/cider/tests/utils/CiderArrowChecker.cpp:326]: yes, duckdb return decimal  

> possible to add some test start from CiderNextgenTestBase?

has one in NextgenCompilerTest

**Comment by xieqi** [comment]
> possible to add some test start from CiderNextgenTestBase?

has one in NextgenCompilerTest

File: src/cider/tests/processor/CiderBatchProcessorTest.cpp [hunk 6]
```diff
@@ -82,6 +82,44 @@ TEST(CiderBatchProcessorTest, statelessProcessorProcessNextBatchTest) {
...
+
+  processor->finish();
+
+  struct ArrowArray output_array;
+  struct ArrowSchema output_schema;
+  processor->getResult(output_array, output_schema);
+
+  EXPECT_EQ(output_array.length, 1);
+  EXPECT_EQ(output_array.n_children, 2);
+  EXPECT_EQ(*(int64_t*)(output_array.children[0]->buffers[1]), 22 * 2);
```
[the remark this answers is not in the document]
Comment (xieqi) [reply src/cider/tests/processor/CiderBatchProcessorTest.cpp:119]: the first is null

File: src/cider/tests/utils/CiderArrowChecker.cpp (hunk 5 above)
Comment (yma11) [reply src/cider/tests/utils/CiderArrowChecker.cpp:326]: okay. This is weird...

File: src/cider/tests/processor/CiderBatchProcessorTest.cpp (hunk 4 above)
Comment (yma11) [reply src/cider/tests/processor/CiderBatchProcessorTest.cpp:109]: equal to do agg on 2 batches data.

File: src/cider/exec/processor/StatefulProcessor.h [hunk 7]
```diff
@@ -22,18 +22,22 @@
...
+  StatefulProcessor(const plan::SubstraitPlanPtr& plan,
+                    const BatchProcessorContextPtr& context,
+                    const cider::exec::nextgen::context::CodegenOptions& codegen_options);
 
   void getResult(struct ArrowArray& array, struct ArrowSchema& schema) override;
 
   Type getProcessorType() const override { return Type::kStateful; };
+
+ private:
+  bool is_groupby_{false};
```
Comment (yma11) [inline src/cider/exec/processor/StatefulProcessor.h:40]: Better to rename to `has_groupby_` too?

File: src/cider/tests/processor/CiderBatchProcessorTest.cpp (hunk 4 above)
Comment (xieqi) [reply src/cider/tests/processor/CiderBatchProcessorTest.cpp:109]: just to test multi batch for agg

**Review by yma11 — APPROVED** [review]

**Pull request merged**